File: frontend/lib/api/browse.ts
/**
 * Browse API Client
 * Handles all browse marketplace API requests
 */

import apiClient from "./client";
import { ContentType, ReviewType } from "./reviews";

// Expert review tier type matching backend enum
export type ExpertReviewTier = "quick" | "standard" | "deep";

// Feedback priority type matching backend enum
export type FeedbackPriority = "validation" | "specific_fixes" | "comprehensive" | "improvement";

// Browse review item with extended metadata
export interface BrowseReviewItem {
  id: number;
  title: string;
  description: string;
  content_type: ContentType;
  review_type: ReviewType;
  status: string;
  created_at: string;
  updated_at: string;
  deadline?: string;
  price?: number;
  currency?: string;
  skills?: string[];
  preview_image?: string;  // URL to preview image (from backend)
  creator_username?: string;
  creator_rating?: number;
  is_featured?: boolean;
  urgency?: "low" | "medium" | "high";
  reviews_requested?: number; // Number of reviews requested (1-10)
  reviews_claimed?: number; // Number of reviews claimed by reviewers
  available_slots?: number; // Computed: reviews_requested - reviews_claimed
  slot_id?: number; // Slot ID for claiming (if available)

  // Skill match score (when user skills provided)
  match_score?: number; // Match percentage (0-100) based on user's skills vs review's skills_needed

  // Expert review tier fields (null for free reviews)
  tier?: ExpertReviewTier; // Expert review tier: quick (5-10min), standard (15-20min), deep (30+ min)
  feedback_priority?: FeedbackPriority; // Primary focus area for the review
  specific_questions?: string[]; // Specific questions the requester wants answered
  context?: string; // Additional context about the project
  estimated_duration?: number; // Estimated review duration in minutes
}

// Browse response with pagination
export interface BrowseResponse {
  items: BrowseReviewItem[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

// Browse request parameters
export interface BrowseParams {
  content_type?: ContentType | "all";
  review_type?: ReviewType | "all";
  sort_by?: "recent" | "price_high" | "price_low" | "deadline";
  deadline?: "urgent" | "week" | "month";
  limit?: number;
  offset?: number;
  search?: string;
  user_skills?: string[]; // User's skills for personalized match scoring
}

/**
 * Get browse reviews from the marketplace
 * Public endpoint - no authentication required
 */
export async function getBrowseReviews(params?: BrowseParams): Promise<BrowseResponse> {
  // Build query string from params
  const queryParams = new URLSearchParams();

  if (params?.content_type && params.content_type !== "all") {
    queryParams.append("content_type", params.content_type);
  }

  if (params?.review_type && params.review_type !== "all") {
    queryParams.append("review_type", params.review_type);
  }

  if (params?.sort_by) {
    queryParams.append("sort_by", params.sort_by);
  }

  if (params?.deadline) {
    queryParams.append("deadline", params.deadline);
  }

  if (params?.search) {
    queryParams.append("search", params.search);
  }

  if (params?.limit) {
    queryParams.append("limit", params.limit.toString());
  }

  if (params?.offset) {
    queryParams.append("offset", params.offset.toString());
  }

  if (params?.user_skills && params.user_skills.length > 0) {
    queryParams.append("user_skills", params.user_skills.join(","));
  }

  const queryString = queryParams.toString();
  const endpoint = `/reviews/browse${queryString ? `?${queryString}` : ""}`;

  // Backend returns { reviews, total, limit, offset }
  // Map to { items, total, limit, offset, has_more }
  const response = await apiClient.get<{ reviews: BrowseReviewItem[]; total: number; limit: number; offset: number }>(endpoint);

  return {
    items: response.reviews || [],
    total: response.total || 0,
    limit: response.limit || 50,
    offset: response.offset || 0,
    has_more: (response.offset || 0) + (response.reviews?.length || 0) < (response.total || 0),
  };
}

/**
 * Get a single review from browse (for detail view)
 */
export async function getBrowseReview(id: number): Promise<BrowseReviewItem> {
  return apiClient.get<BrowseReviewItem>(`/reviews/browse/${id}`);
}

/**
 * Claim response from backend
 */
export interface ClaimReviewResponse {
  success: boolean;
  message: string;
  review_request_id: number;
  slot_id: number;  // The newly created/claimed slot ID
  reviews_claimed: number;
  available_slots: number;
  is_fully_claimed: boolean;
}

/**
 * Claim a review slot (requires authentication)
 * For reviews with multiple slots, claims one available slot
 * Returns the slot_id which is needed to redirect to the review writing page
 */
export async function claimReviewSlot(id: number): Promise<ClaimReviewResponse> {
  return apiClient.post<ClaimReviewResponse>(`/reviews/${id}/claim`);
}

File: backend/app/schemas/browse.py
"""Browse-related Pydantic schemas for public marketplace"""

import enum
from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field

from app.models.review_request import (
    ContentType,
    ReviewType,
    ReviewStatus,
    ReviewTier,
    FeedbackPriority
)


class SortOption(str, enum.Enum):
    """Sort options for browse endpoint"""
    RECENT = "recent"  # created_at DESC
    PRICE_HIGH = "price_high"  # budget DESC
    PRICE_LOW = "price_low"  # budget ASC
    DEADLINE = "deadline"  # deadline ASC (closest deadline first)


class DeadlineFilter(str, enum.Enum):
    """Deadline urgency filters"""
    URGENT = "urgent"  # Less than 24 hours
    THIS_WEEK = "this_week"  # Less than 7 days
    THIS_MONTH = "this_month"  # Less than 30 days
    FLEXIBLE = "flexible"  # More than 30 days or no deadline


class UrgencyLevel(str, enum.Enum):
    """Urgency level calculated from deadline"""
    URGENT = "urgent"  # <24h
    THIS_WEEK = "this_week"  # <7d
    THIS_MONTH = "this_month"  # <30d
    FLEXIBLE = "flexible"  # >30d or None


# ===== Creator Info Schema =====

class CreatorInfo(BaseModel):
    """
    Public creator information (no sensitive data).

    Only includes information safe to display publicly:
    - User ID (for linking to public profile)
    - Full name (for display)
    - Avatar URL (for visual identification)

    NEVER includes: email, phone, payment info, private data
    """
    id: int = Field(..., description="User ID")
    full_name: Optional[str] = Field(None, description="Creator's display name")
    avatar_url: Optional[str] = Field(None, description="Creator's avatar image URL")

    class Config:
        from_attributes = True


# ===== Browse Review Schema =====

class BrowseReviewItem(BaseModel):
    """
    Public review request data for marketplace browsing.

    Contains all information needed to display a review in the marketplace:
    - Basic review details (title, description, type)
    - Pricing information for expert reviews
    - Deadline and urgency
    - Creator information (public only)
    - Preview image/thumbnail
    - Skills needed
    - Multi-review claim status
    - Expert review tier information
    """
    id: int = Field(..., description="Review request ID")
    title: str = Field(..., description="Review title")
    description: str = Field(..., description="Review description")
    content_type: ContentType = Field(..., description="Type of content to review")
    review_type: ReviewType = Field(..., description="Review service type (free/expert)")
    price: Optional[Decimal] = Field(
        None,
        description="Price in USD for expert reviews (null for free reviews)",
        decimal_places=2
    )
    deadline: Optional[datetime] = Field(None, description="Review deadline (ISO 8601)")
    status: ReviewStatus = Field(..., description="Current review status")
    created_at: datetime = Field(..., description="When review was created (ISO 8601)")
    creator: CreatorInfo = Field(..., description="Public creator information")
    preview_image: Optional[str] = Field(
        None,
        description="Preview image/thumbnail URL (first file if available)"
    )
    skills_needed: List[str] = Field(
        default_factory=list,
        description="Skills or areas needed for review (parsed from feedback_areas)"
    )
    urgency: UrgencyLevel = Field(..., description="Deadline urgency level")

    # Multi-review support
    reviews_requested: int = Field(..., ge=1, le=10, description="Total number of reviews requested")
    reviews_claimed: int = Field(..., ge=0, description="Number of review slots claimed")
    available_slots: int = Field(..., ge=0, description="Number of available review slots")

    # Skill match score (when user skills provided)
    match_score: Optional[int] = Field(
        None,
        ge=0,
        le=100,
        description="Skill match percentage (0-100) based on user's skills vs review's skills_needed. Null if no user skills provided."
    )

    # Expert review tier fields (null for free reviews)
    tier: Optional[ReviewTier] = Field(
        None,
        description="Expert review tier: quick (5-10min), standard (15-20min), deep (30+ min). Null for free reviews."
    )
    feedback_priority: Optional[FeedbackPriority] = Field(
        None,
        description="Primary focus area for the review"
    )
    specific_questions: Optional[List[str]] = Field(
        None,
        description="Specific questions the requester wants answered"
    )
    context: Optional[str] = Field(
        None,
        description="Additional context about the project"
    )
    estimated_duration: Optional[int] = Field(
        None,
        description="Estimated review duration in minutes"
    )

    class Config:
        from_attributes = True


# ===== Claim Review Schema =====

class ClaimReviewSlotRequest(BaseModel):
    """Request schema for claiming a review slot"""
    review_request_id: int = Field(..., gt=0, description="ID of the review request to claim")


class ClaimReviewSlotResponse(BaseModel):
    """Response schema after claiming a review slot"""
    success: bool = Field(..., description="Whether the claim was successful")
    message: str = Field(..., description="Status message")
    review_request_id: int = Field(..., description="ID of the claimed review request")
    slot_id: int = Field(..., description="ID of the claimed review slot")
    reviews_claimed: int = Field(..., ge=0, description="Total slots now claimed")
    available_slots: int = Field(..., ge=0, description="Remaining available slots")
    is_fully_claimed: bool = Field(..., description="Whether all slots are now claimed")


# ===== Browse Response Schema =====

class BrowseReviewsResponse(BaseModel):
    """
    Response schema for browse endpoint with pagination.

    Provides:
    - List of reviews matching filters
    - Total count for pagination
    - Current limit and offset
    """
    reviews: List[BrowseReviewItem] = Field(
        default_factory=list,
        description="List of review requests"
    )
    total: int = Field(..., description="Total number of matching reviews", ge=0)
    limit: int = Field(..., description="Number of results per page", ge=1, le=100)
    offset: int = Field(..., description="Current pagination offset", ge=0)

    @property
    def has_more(self) -> bool:
        """Check if there are more results available"""
        return (self.offset + len(self.reviews)) < self.total

    @property
    def next_offset(self) -> Optional[int]:
        """Get the offset for the next page, or None if no more results"""
        if self.has_more:
            return self.offset + self.limit
        return None

    @property
    def previous_offset(self) -> Optional[int]:
        """Get the offset for the previous page, or None if on first page"""
        if self.offset > 0:
            return max(0, self.offset - self.limit)
        return None

    class Config:
        from_attributes = True
